alg/los.cpp

```diff
@@ -0,0 +1,232 @@
+                           const int y, const double z)
+{
+    double terrainHeight;
+    if (GetElevation(hBand, x, y, terrainHeight))
+    {
+        return z > terrainHeight;
+    }
+    else
+    {
+        return false;
+    }
+}
+
+/************************************************************************/
+/*                        GDALIsLineOfSightVisible()                    */
+/************************************************************************/
+
+/**
+ * Check Line of Sight between two points.
+ * Both input coordinates must be within the raster coordinate bounds.
+ *
+ * This algorithm will check line of sight using a Bresenham algorithm.
+ * https://www.researchgate.net/publication/2411280_Efficient_Line-of-Sight_Algorithms_for_Real_Terrain_Data
+ * Line of sight is computed in raster coordinate space, and thus may not be appropriate.
+ * For example, datasets referenced against geographic coordinate at high latitudes may have issues.
+ *
+ * @param hBand The band to read the DEM data from. This must NOT be null.
+ * 
+ * @param xA The X location (raster column) of the first point to check on the raster.
+ *
+ * @param yA The Y location (raster row) of the first point to check on the raster.
+ * 
+ * @param zA The Z location (height) of the first point to check.
+ * 
+ * @param xB The X location (raster column) of the second point to check on the raster.
+ *
+ * @param yB The Y location (raster row) of the second point to check on the raster.
+ * 
+ * @param zB The Z location (height) of the second point to check.
+ * 
+ * @param papszOptions Options for the line of sight algorithm (currently ignored).
+ * 
+ * @return True if the two points are within Line of Sight.
+ *
+ * @since GDAL 3.9
+ */
+
+bool GDALIsLineOfSightVisible(const GDALRasterBandH hBand, const int xA,
+                              const int yA, const double zA, const int xB,
+                              const int yB, const double zB,
+                              CPL_UNUSED CSLConstList papszOptions)
+{
+    VALIDATE_POINTER1(hBand, "GDALIsLineOfSightVisible", false);
+
+    // Perform a preliminary check of the start and end points.
+    if (!IsAboveTerrain(hBand, xA, yA, zA))
+    {
+        return false;
+    }
+    if (!IsAboveTerrain(hBand, xB, yB, zB))
+    {
+        return false;
+    }
+
+    // If both X and Y are the same, no further checks are needed.
+    if (xA == xB && yA == yB)
+    {
+        return true;
+    }
+
+    // TODO if both X's or Y's are the same, it could be optimized for vertical/horizontal lines.
+
+    // Use an interpolated Z height with 2D bresenham for the remaining cases.
+
+    // Lambda for Linear interpolate like C++20 std::lerp.
+    auto lerp = [](const double a, const double b, const double t)
+    { return a + t * (b - a); };
+
+    // Lambda for getting Z test height given x input along the bresenham line.
+    auto GetZValue = [&](const int x) -> double
```

Comment:
GetZValue with just the "x" argument will not work for LoS computations where x1 == x2. You'd likely need to pass (x,y), and compute the ratio ``sqrt((SQUARE(x - x1) + SQUARE(y - y1)) / (SQUARE(x2 - x1) + SQUARE(y2 - y1)))``

Reply:
Good good. I'll add a test for a vertical and horizontal line. 

Also, exact vertical or horizontal line is a special case that can be optimized (don't call bresenham, just iterate directly).

Reply:
> Also, exact vertical or horizontal line is a special case

but my point would also be valid for an almost vertical line (that is when the ratio abs((x2-x1)/(y2-y1)) is low.
An alternative implementation could be to determine which of abs(x2-x1) and abs(y2-y1) is the greater, and compute the ration according to x if dx > dy, and y otherwise.

Reply:
Fixed in b9ebc37210b693c3f449dae659a8143af04510f5, and I added tests for vertical/horizontal cases in 248e0de42b2c9c872300465c5defdc0cf99c0b73.

Reply:
Using sqrt operators over and over in this manner sort of defeats the purpose of using bresenham, but I want to get benchmarking set up first before optimizing, and not in this PR if that's ok.